### PR TITLE
Node: Safe use of `document` for preview

### PR DESCRIPTION
**Pull request by DylanPiercey**
## What I did

Avoid referencing `document.location` when it's possible for it not to be there.

I am currently using the following apis in order to reimplement the testing portion for [Marko Storybook](https://github.com/storybookjs/marko/blob/next/testing/README.md) which I am upgrading to support storybook 7.

```js
import {
  composeStories,
  composeStory,
  setProjectAnnotations,
} from "@storybook/preview-api";
```

These defer the implementation of the story composition for testing which is great!

The issue I ran into is that the `@storybook/preview-api` uses a browser api which does not exist in server environments and for Marko apps (likely Svelte, Solid and anything else with a different ssr compilation) it is nice to also test these fixtures in a server context.

With the simple change in this PR it seems this API works fine in a bare node environment and there is already [precedent in the storybook codebase for checking if the `document.location` exists](https://github.com/storybookjs/storybook/blob/next/code/lib/preview-api/src/modules/client-api/queryparams.ts#L7).

### Testing

I'm not exactly sure how best to add tests for this but am open to ideas.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Use any storybook project where the stories could be rendered on the server.
2. Import the stories, and the `composeStory` api in a node script.
3. Try and render the story using the `composeStory` api.

-->

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

**Comment by DylanPiercey** [comment]
I found the main issue is calling `@storybook/preview-api#start`. So another work around for me is to just avoid that in the testing setups.

I'm happy with that work around, but I'll leave up the PR if you all wish to discuss.

File: code/lib/preview-api/src/modules/preview-web/UrlStore.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ const getQueryString = ({
   selection?: Selection;
   extraParams?: qs.ParsedQs;
 }) => {
-  const { search = '' } = document.location;
+  const search = (document && document.location && document.location.search) || '';
```
Comment (ndelangen) [inline code/lib/preview-api/src/modules/preview-web/UrlStore.ts:25]: I'm ok with this defensive programming, just to be on the safe side.

Minor nit pick: could you use optional chaining here to keep the code concise?

@shilman WDYT?

Comment (DylanPiercey) [reply code/lib/preview-api/src/modules/preview-web/UrlStore.ts:25]: Happy to use optional chaining, I only avoided it to stick with the pattern used here: https://github.com/storybookjs/storybook/blob/next/code/lib/preview-api/src/modules/client-api/queryparams.ts#L7

Comment (ndelangen) [reply code/lib/preview-api/src/modules/preview-web/UrlStore.ts:25]: We should start using optional chaining there too 😅

**Comment by tmeasday** [comment]
@DylanPiercey I'm confused as to why that code is executing if you are only using the listed APIs?

**Comment by yannbf** [comment]
> @DylanPiercey I'm confused as to why that code is executing if you are only using the listed APIs?

This happens when APIs like `composeStories` are called. We had a similar issue when trying `composeStories` with Playwright CT. This PR seems like it would fix those problems!

File: code/lib/preview-api/src/modules/preview-web/UrlStore.ts (hunk 1 above)
Comment (DylanPiercey) [reply code/lib/preview-api/src/modules/preview-web/UrlStore.ts:25]: @ndelangen do you think it's worth changing this? Or create a separate issue for switching stuff to optional chaining more broadly?

**Comment by tmeasday** [comment]
> This happens when APIs like composeStories are called in node, in non JSDOM environments. We had a similar issue when trying composeStories with Playwright CT. 

What does `composeStories` call that ends up calling these lines?

> This PR seems like it would fix those problems!

Agree, but it seems like maybe a bandaid. I don't think a preview should be getting initiated if there is no "preview", at least naively (maybe there's a good reason it needs to). Can we figure out if it absolutely necessary for these lines to execute?

**Comment by DylanPiercey** [comment]
@tmeasday to be clear calling `composeStory` or `composeStories` does not cause these lines to execute. What does is calling the `start` api from preview. Since in my case start was called unconditionally and the test apis were exposed through the same module this is more of a side effect.

I changed it to lazily call start so that it doesn't happen in testing, which is a good enough work around for me, but maybe this change is still worth while?

**Comment by tmeasday** [comment]
> What does is calling the start api from preview. Since in my case start was called unconditionally and the test apis were exposed through the same module this is more of a side effect.

There's no reason to call `start` outside of story store v6 which is definitely not relevant in this scenario. We'll be removing that code in SB8.0 in any case.

> I changed it to lazily call start so that it doesn't happen in testing, which is a good enough work around for me, but maybe this change is still worth while?

Can you explain this to me / link me to this code?

**Review by ndelangen — APPROVED** [review]

**Comment by tmeasday** [comment]
@ndelangen although this PR isn't causing any harm, I'm still unsure why it's necessary. Plus I wonder if we need to check `typeof document !== 'undefined'` rather than just `!!document`, TBH.

**Comment by ndelangen** [comment]
I agree @tmeasday

**Comment by ndelangen** [comment]
@yannbf considering you've worked a lot on composable stories.. Do you think this PR should still be merged?

**Comment by yannbf** [comment]
> @yannbf considering you've worked a lot on composable stories.. Do you think this PR should still be merged?

I don't think these changes are much needed anymore because we don't instantiate PreviewWeb like we used to in Storybook 7, which was part of exporting clientApi. However I think it's fine to merge them, so I'd say let's do it! But would be nice to change the assertions to use optional chaining if possible

**Review by ndelangen — APPROVED** [review]

**Pull request merged**